### PR TITLE
interpreter: Initial WGPU interpreter.

### PR DESCRIPTION
**Pull request by PapyChacal**
Here's a first draft/example for a WebGPU-based interpreter for the GPU dialect, including JIT-compilation through WGSL.
I only implement very specific cases of `gpu.func`, `gpu.launch_func`, `gpu.alloc` and `gpu.memcpy` here, to end up with a bearable PR!

NB: contains a minor fix on `gpu.memcpy`, this wasn't an issue anywhere else because the constructor swapped the argument again, but my first definition of this operation was wrong, as the destination is listed *before* the source for this op, as opposed to `memref.copy`.

This requires the addition of Vulkan drivers in the CI

### CODE REVIEW

**Comment by PapyChacal** [comment]
Oops, forgot my denial of thys Pyright mess: the wgpu-py package really doesn't play nice with typing at time of writing. Soyunds like a nice PR to do there, but I'll try to fill the holes myself here first

**Comment by PapyChacal** [comment]
Okay, so I'm blocked on this formatting issue, I think my local ruff disagree with the CI one? :upside_down_face: 

More importantly, the test fail because WGPU doesn't find a device to initialize on the CI runner. I'll see if I can fix this, meanwhile feel free to play around locally!

**Review by cigarichard — APPROVED** [review]

File: tests/interpreters/test_wgpu.py
Comment (superlopuh) [inline tests/interpreters/test_wgpu.py:82]: A) I didn't know this was possible, pretty cool
B) The interpreter has an optional StringIO that the interpreter functions should use for this purpose

File: xdsl/interpreters/experimental/wgpu.py
Comment (superlopuh) [inline xdsl/interpreters/experimental/wgpu.py:94]:
```suggestion
        if args or op.asyncToken:
```

File: xdsl/interpreters/experimental/wgpu.py
Comment (superlopuh) [inline xdsl/interpreters/experimental/wgpu.py:155]:
```suggestion
        if op.asyncToken is not None or op.asyncDependencies:
```

File: xdsl/interpreters/experimental/wgpu.py
Comment (superlopuh) [inline xdsl/interpreters/experimental/wgpu.py:172]: Looks like we don't really need the list construction, does this work?
```suggestion
        dispatch = gridSize
```

**Review by superlopuh — CHANGES_REQUESTED** [review]
Exciting, LGTM modulo small comments

File: tests/interpreters/test_wgpu.py
Comment (PapyChacal) [reply tests/interpreters/test_wgpu.py:82]: This is copy pasta from other tests of ours 🙂
Will see if I can use this other parameter

File: xdsl/interpreters/experimental/wgpu.py
Comment (PapyChacal) [reply xdsl/interpreters/experimental/wgpu.py:172]: Nope, tuple is immutable

**Review by superlopuh — APPROVED** [review]

File: .github/workflows/ci-mlir.yml
Comment (webmiche) [inline .github/workflows/ci-mlir.yml:33]: Personally, I would prefer to separate our testing environments a bit. Otherwise, we might run into versioning issues were certain packages are required, but not built with just the default config.

For that reason, I think I would have prefered this to be its own separate CI run.

What do others think? I guess our requirements-optional.txt and what is needed where is a big mess already, so this might not just be an issue with these tests, but something deeper. tagging @math-fehr @superlopuh and @PapyChacal here :)

Comment (PapyChacal) [reply .github/workflows/ci-mlir.yml:33]: I would love to look into structuring those things for sure, but probably later I. August as we all have deadlines running currently 🙂

Comment (superlopuh) [reply .github/workflows/ci-mlir.yml:33]: I would be interested in splitting the requirements-optional per subproject, with `extras` installing them all.

**Pull request merged**
